dosbox-staging/dosbox-staging
# Rework translated messages handling

# Description

Rewritten translated messages handling; now has a format string sanitizer (already used by me to create PR https://github.com/dosbox-staging/dosbox-staging/pull/3964) to prevent crashes due to outdated translations, `MSG_*` functions now take `std::string` arguments, plus there are several user-visible improvements listed in the proposed _Release notes_.

Limitation: the translation fallback to English (when the code page does not provide needed Cyrillic / Greek / etc. script) does not work in the generated `Z:\AUTOEXEC.BAT` - we can't risk string length changes when it s being executed. Maybe I'll work it around at some point, but I don't want to do this now.

BTW. The rework in `messages.cpp`  is a precondition for switching translations at runtime.


## Related issues

Fixes https://github.com/dosbox-staging/dosbox-staging/issues/3498 (together with previous localisation PRs)


# Release notes

- Added `--list-code-pages` command line option, to list all the available code pages with a brief description.
- The `KEYB` command output is now event more informative, it now prints code page description.
- Non-Latin translations are only used for DOS messages if the selected code page supports the given script; otherwise DOSBox reverts to English messages,


# Manual testing

- start DOSBox with `language = fr` setting - it will display some warnings about several translated messages not matching the English strings regarding the format strings (`fr` and `ru` translations weren't updated since a long time...) - 'time /?' command will now print a partially translated help string, as the non-matching translated string will be rejected by the format string sanitizer
- start DOSBox with `--list-code-pages` command line argument - it should display information about all the bundled code pages
- withing DOSBox try commands `keyb pl`, `keyb us 113`, etc. - command should now display code page descriptions

For testing modified language handling, I suggest to use this dummy Greek translation - rename it to `gr-dummy.lng`, put into the build directory where other `*.lng` files are stored, and set `language = gr-dummy` in the config file:

[gr-dummy.txt](https://github.com/user-attachments/files/17951748/gr-dummy.txt)

When you use this dummy language:
- strings in the titlebar (`cycles/ms`, etc.) should be replaced with some dummy Greek letters, all the time
- `dosbox --list-code-pages`, `dosbox --list-layouts` and `dosbox --list-countries` commands will display Greek letters from the dummy translation file no matter what you put into `keyboard_layout` config option (BTW, Greek keyboard layout is `gk`, `gk220`, or `gk459`)
- inside DOSBox, you can select a Greek code page (`keyb us 869` will do so without setting the non-standard keyboard layout, `keyb us` restores a Latin-only code page) - an illegal command error (just type something dummy in the command prompt) prints out some Greek letters if and only if a Greek code page is set
- similarly, `keyb /list` prints out some Greek letters at the start of the message if and only if a Greek code page is set
- each time you select a non-Greek code page, a warning should be logged


# Checklist

I have:
